File: src/Jump_Game_II.py
# -*- coding: utf-8 -*-


class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def jump(self, nums):
        upper_bound, rec_bound, cnt = 0, 0, 0
        for i in range(len(nums) - 1):
            upper_bound = max(upper_bound, i + nums[i])
            if i == rec_bound:
                cnt += 1
                rec_bound = upper_bound
        return cnt


if __name__ == '__main__':
    solution = Solution()
    print solution.jump([2, 3, 1, 1, 4])

File: src/algorithms/cpp/Jump_Game_II.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <stack>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int jump(vector<int>& nums) {
        int upper_bound = 0, rec_bound = 0, cnt = 0;
        for (int i = 0; i < SZ(nums) - 1; ++ i) {
            get_max(upper_bound, i + nums[i]);
            if (i == rec_bound) {
                rec_bound = upper_bound;
                cnt += 1;
            }
        }
        return cnt;
    }
};

int main() {
    return 0;
}
